Implement declarativeNetRequest.setExtensionActionOptions

b-weinstein opened the pull request:
#### c26e66feb2aa3ac2cb62f6a15ea66c0dacef644e
<pre>
Implement declarativeNetRequest.setExtensionActionOptions
<a href="https://bugs.webkit.org/show_bug.cgi?id=265829">https://bugs.webkit.org/show_bug.cgi?id=265829</a>
<a href="https://rdar.apple.com/118476776">rdar://118476776</a>

Reviewed by Timothy Hatcher.

This API has two use cases:
1) To opt an extension into a behavior where the extension&apos;s action shows the number of blocked resources on the current page
2) To manually increment or decrement this badged number

It&apos;s a bit unfortunate that these two disparate behaviors are combined into one API, but here we are. This PR implements both of
them and adds tests for them.

* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::NavigationClient::contentRuleListNotification): Call into the WebExtensionController.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
(WebKit::WebExtensionContext::shouldDisplayBlockedResourceCountAsBadgeText): Returns whether or not the blocked resource count should be the badge text.
(WebKit::WebExtensionContext::saveShouldDisplayBlockedResourceCountAsBadgeText): Sets whether or not the blocked resource count should be the badge
text, and saves to disk.
(WebKit::WebExtensionContext::incrementActionCountForTab): Get the action for the tab and increment the blocked resource count.
(WebKit::WebExtensionContext::declarativeNetRequestDisplayActionCountAsBadgeText): Call saveShouldDisplayBlockedResourceCountAsBadgeText with the new value, and
clear any blocked resource counts if the flag is turned off.
(WebKit::WebExtensionContext::declarativeNetRequestIncrementActionCount): Call into incrementActionCountForTab after finding the tab.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::clearCustomizations): Clear the blocked resource count.
(WebKit::WebExtensionAction::clearBlockedResourceCount): Ditto.
(WebKit::WebExtensionAction::badgeText const): If we have a blocked resource count - use it as the badge text.
(WebKit::WebExtensionAction::incrementBlockedResourceCount): Modify the blocked resource count member variable.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::handleContentRuleListNotification): Iterate over all of the actions, find the extension if it exists, and call incrementActionCountForTab.
* Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
(WebKit::WebExtensionAPIDeclarativeNetRequest::setExtensionActionOptions): Perform object validation and call into the UI process based on if flavor (1) or (2) was called.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/271561@main">https://commits.webkit.org/271561@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dfb2ff0925be7b8a927722670986d1640cdef812

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28799 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7442 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30172 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31418 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26275 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29369 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9576 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4801 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29069 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6172 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/24761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/5357 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/25771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32757 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26369 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/26212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/31748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5478 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/3649 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7099 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6885 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5950 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5996 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -173,6 +174,54 @@
+    writeStateToStorage();
+}
+
+void WebExtensionContext::incrementActionCountForTab(WebExtensionTab* tab, ssize_t incrementAmount)
```
`WebExtensionTab&` would be best, since tab is required right? if tab is null it will increment the default action, for all tabs, which would be unexpected for this API I think.

b-weinstein replied:
Tab is required, yes.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -173,6 +174,54 @@
+        return;
+    }
+
+    incrementActionCountForTab(tab.get(), (int)increment);
```
`ssize_t`, if a cast is even needed.

b-weinstein replied:
Whoops, I meant to get rid of that.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
```diff
@@ -417,6 +428,16 @@ - (void)invalidateIntrinsicContentSize
+void WebExtensionAction::incrementBlockedResourceCount(ssize_t amount)
+{
+    if (m_blockedResourceCount + amount < 0)
+        return;
```
Should this still take the count to zero? What do other browsers do?

b-weinstein replied:
I think it should be able to take it to zero - that would clear the badge text.

xeenon replied:
I mean, if `m_blockedResourceCount` is 1, and `amount` is -2, should it just reset `m_blockedResourceCount` to 0? Right now it will stay at 1 with the return.

b-weinstein replied:
I will see what Chrome does.

b-weinstein replied:
What you described is what Chrome does. I will make that change.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -294,6 +295,25 @@
+            if (context->uniqueIdentifier() != contentRuleListIdentifier)
+                continue;
+
+            RefPtr tab = context->getTab(pageID);
+            if (!tab)
+                break;
```
You should move this out of the loops, since it is the same for all results.

b-weinstein replied:
Is it? Getting the tab seems to require the context, and what happens if one specific extension doesn't have access to a tab?

If they are the same, then I'm fine moving this out of the loop - how would you recommend getting the tab for one particular extension?

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
```diff
@@ -116,6 +119,7 @@ class WebExtensionAction : public API::ObjectImpl<API::Object::Type::WebExtensio
     RetainPtr<NSDictionary> m_customIcons;
     String m_customLabel;
     String m_customBadgeText;
+    ssize_t m_blockedResourceCount;
```
`ssize_t m_blockedResourceCount { 0 };`

xeenon commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/xcshareddata/xcschemes/TestWebKitAPI.xcscheme:
```diff
@@ -64,6 +64,12 @@
+            argument = ""
+            isEnabled = "NO">
+         </CommandLineArgument>
+      </CommandLineArguments>
```
Revert.

b-weinstein replied:
Whoops I thought I had.

b-weinstein committed:
Implement declarativeNetRequest.setExtensionActionOptions
https://bugs.webkit.org/show_bug.cgi?id=265829
rdar://118476776

Reviewed by Timothy Hatcher.

This API has two use cases:
1) To opt an extension into a behavior where the extension's action shows the number of blocked resources on the current page
2) To manually increment or decrement this badged number

It's a bit unfortunate that these two disparate behaviors are combined into one API, but here we are. This PR implements both of
them and adds tests for them.

* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::NavigationClient::contentRuleListNotification): Call into the WebExtensionController.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
(WebKit::WebExtensionContext::shouldDisplayBlockedResourceCountAsBadgeText): Returns whether or not the blocked resource count should be the badge text.
(WebKit::WebExtensionContext::saveShouldDisplayBlockedResourceCountAsBadgeText): Sets whether or not the blocked resource count should be the badge
text, and saves to disk.
(WebKit::WebExtensionContext::incrementActionCountForTab): Get the action for the tab and increment the blocked resource count.
(WebKit::WebExtensionContext::declarativeNetRequestDisplayActionCountAsBadgeText): Call saveShouldDisplayBlockedResourceCountAsBadgeText with the new value, and
clear any blocked resource counts if the flag is turned off.
(WebKit::WebExtensionContext::declarativeNetRequestIncrementActionCount): Call into incrementActionCountForTab after finding the tab.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::clearCustomizations): Clear the blocked resource count.
(WebKit::WebExtensionAction::clearBlockedResourceCount): Ditto.
(WebKit::WebExtensionAction::badgeText const): If we have a blocked resource count - use it as the badge text.
(WebKit::WebExtensionAction::incrementBlockedResourceCount): Modify the blocked resource count member variable.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::handleContentRuleListNotification): Iterate over all of the actions, find the extension if it exists, and call incrementActionCountForTab.
* Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
(WebKit::WebExtensionAPIDeclarativeNetRequest::setExtensionActionOptions): Perform object validation and call into the UI process based on if flavor (1) or (2) was called.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/271561@main
